x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java

```diff
@@ -85,31 +96,83 @@ public boolean isSafeAction() {
-            ShrunkenIndexCheckStep.NAME);
+        CopyExecutionStateStep copyMetadata = new CopyExecutionStateStep(copyMetadataKey, dataStreamCheckBranchingKey,
+            SHRUNKEN_INDEX_PREFIX, ShrunkenIndexCheckStep.NAME);
+        // by the time we get to this step we have 2 indices, the source and the shrunken one. we now need to choose an index
+        // swapping strategy such that the shrunken index takes the place of the source index (which is also deleted).
+        // if the source index is part of a data stream it's a matter of replacing it with the shrunken index one in the data stream and
+        // then deleting the source index; otherwise we'll use the alias management api to atomically transfer the aliases from source to
+        // the shrunken index and delete the source
+        BranchingStep isDataStreamBranchingStep = new BranchingStep(dataStreamCheckBranchingKey, aliasKey, replaceDataStreamIndexKey,
+            (index, clusterState) -> {
+                IndexAbstraction indexAbstraction = clusterState.metadata().getIndicesLookup().get(index.getName());
+                assert indexAbstraction != null : "invalid cluster metadata. index [" + index.getName() + "] was not found";
+                return indexAbstraction.getParentDataStream() != null;
+            });
         ShrinkSetAliasStep aliasSwapAndDelete = new ShrinkSetAliasStep(aliasKey, isShrunkIndexKey, client, SHRUNKEN_INDEX_PREFIX);
+        ReplaceDataStreamBackingIndexStep replaceDataStreamBackingIndex = new ReplaceDataStreamBackingIndexStep(replaceDataStreamIndexKey,
+            deleteIndexKey, SHRUNKEN_INDEX_PREFIX);
+        DeleteStep deleteSourceIndexStep = new DeleteStep(deleteIndexKey, isShrunkIndexKey, client);
         ShrunkenIndexCheckStep waitOnShrinkTakeover = new ShrunkenIndexCheckStep(isShrunkIndexKey, nextStepKey, SHRUNKEN_INDEX_PREFIX);
         return Arrays.asList(conditionalSkipShrinkStep, waitForNoFollowersStep, readOnlyStep, setSingleNodeStep, checkShrinkReadyStep,
-            shrink, allocated, copyMetadata, aliasSwapAndDelete, waitOnShrinkTakeover);
+            shrink, allocated, copyMetadata, isDataStreamBranchingStep, aliasSwapAndDelete, waitOnShrinkTakeover,
+            replaceDataStreamBackingIndex, deleteSourceIndexStep);
+    }
+
+    static BiPredicate<Index, ClusterState> getSkipShrinkStepPredicate(int targetNumberOfShards) {
+        return (index, clusterState) -> {
+            IndexMetadata indexMetadata = clusterState.getMetadata().index(index);
+            if (indexMetadata == null) {
+                // Index must have been since deleted, skip the shrink action
+                logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", NAME, index.getName());
+                return true;
+            }
+
+            if (indexMetadata.getNumberOfShards() == targetNumberOfShards) {
+                return true;
```

Comment:
Super minor, but I think we should add a debug log that shrink is being skipped because the targeted number of shards has already been met